src/librustdoc/html/render/print_item.rs

```diff
@@ -1492,61 +1496,134 @@ fn item_constant(w: &mut Buffer, cx: &mut Context<'_>, it: &clean::Item, c: &cle
-        render_struct(w, it, Some(&s.generics), s.ctor_kind, &s.fields, "", true, cx);
-    });
+    #[derive(Template)]
+    #[template(path = "item_struct.html")]
+    struct ItemStruct<'a, 'cx> {
+        cx: std::cell::RefCell<&'a mut Context<'cx>>,
+        it: &'a clean::Item,
+        s: &'a clean::Struct,
+    }
 
-    write!(w, "{}", document(cx, it, None, HeadingOffset::H2));
+    impl<'a, 'cx: 'a> ItemStruct<'a, 'cx> {
+        fn render_attributes_in_code<'b>(
+            &'b self,
+        ) -> impl fmt::Display + Captures<'a> + 'b + Captures<'cx> {
+            display_fn(move |f| {
+                let tcx = self.cx.borrow().tcx();
+                write!(f, "{}", render_attributes_in_code(self.it, tcx))
+            })
+        }
 
-    let mut fields = s
-        .fields
-        .iter()
-        .filter_map(|f| match *f.kind {
-            clean::StructFieldItem(ref ty) => Some((f, ty)),
-            _ => None,
-        })
-        .peekable();
-    if let None | Some(CtorKind::Fn) = s.ctor_kind {
-        if fields.peek().is_some() {
-            write!(
-                w,
-                "<h2 id=\"fields\" class=\"fields small-section-header\">\
-                     {}{}<a href=\"#fields\" class=\"anchor\">§</a>\
-                 </h2>\
-                 {}",
-                if s.ctor_kind.is_none() { "Fields" } else { "Tuple Fields" },
-                document_non_exhaustive_header(it),
-                document_non_exhaustive(it)
-            );
-            for (index, (field, ty)) in fields.enumerate() {
+        fn render_struct<'b>(&'b self) -> impl fmt::Display + Captures<'a> + 'b + Captures<'cx> {
+            display_fn(move |f| {
+                let cx = self.cx.borrow();
+                let v = render_struct(
+                    self.it,
+                    Some(&self.s.generics),
+                    self.s.ctor_kind,
+                    &self.s.fields,
+                    "",
+                    true,
+                    *cx,
+                );
+                write!(f, "{v}")
+            })
+        }
+
+        fn document<'b>(&'b self) -> impl fmt::Display + Captures<'a> + 'b + Captures<'cx> {
+            display_fn(move |f| {
+                let mut cx = self.cx.borrow_mut();
+                let v = document(*cx, self.it, None, HeadingOffset::H2);
+                write!(f, "{v}")
+            })
+        }
+
+        fn fields(&self) -> impl Iterator<Item = (&clean::Item, &clean::Type)> {
+            self.s.fields.iter().filter_map(|item| match *item.kind {
+                clean::StructFieldItem(ref ty) => Some((item, ty)),
+                _ => None,
+            })
+        }
+
+        fn should_render_fields(&self) -> bool {
+            matches!(self.s.ctor_kind, None | Some(CtorKind::Fn))
+                && self.fields().peekable().peek().is_some()
+        }
+
+        fn document_non_exhaustive_header(&self) -> &str {
+            document_non_exhaustive_header(self.it)
```

Comment:
This wrapping is really not great. Isn't it possible to call this function directly from the askama?

Reply:
One thing that we can try to explore is implementing something like this [PR](https://github.com/rust-lang/rust/pull/110470/files#diff-0f33e2ad58ad54105ba42ffe18470c6550a6d86c71c4fbd5dcacc3a6c18e1212) to prevent multiple wrappings across multiple items

Reply:
Indeed!

Reply:
I'll try to come up with a solution for that and publish another PR once it's ready since it will involve other items.

What do you think?

Reply:
I'm thinking of using a `derive` macro something like this,

# Usage
```rust
#[derive(ItemTemplate)]
#[template(path = "item_struct.html")]
struct ItemStruct {
    // cx, it, s, (and other fields)
}

impl ItemStruct {
    // implementation of custom methods here
}
```

# Macro Expansion
```rust
impl ItemStruct {
    // implementation of common methods here, e.g. render_assoc_items
}
```

Haven't tested this yet, just dumping my ideas here first. Any opinions?

Reply:
At first, I tried to use plain `trait` but faced some issues,
- Can't return `impl Trait`s in trait definition: this is used for default method implementation of common methods
- The previous PR uses `borrow_mut()` to extract `cx` and `it` out of the struct: this will create a good amount of boilerplate code

Reply:
I'm curious to see that. The only problem with proc-macro is that it would require to create a new `rustdoc-proc-macro` crate. Maybe let's go with the boilerplate code first and merge this first iteration and then we can think about how to reduce it later on?

Reply:
> * Can't return `impl Trait`s in trait definition: this is used for default method implementation of common methods

FYI: "Return position impl Trait in traits" RFC is now entering its final comment period [(comment)](https://github.com/rust-lang/rfcs/pull/3425#issuecomment-1559934038)